File: rust/day1/src/main.rs
use std::fs;
use std::env::args;

fn read_input(str: String) -> Vec<String> {
    let result = fs::read_to_string(str).expect("Failed to read input.txt");
    result.lines().map(|s| s.to_string()).collect()
}

fn split_string(input: Vec<String>) -> [Vec<i64>; 2] {
    let mut first_column: Vec<i64> = Vec::new();
    let mut second_column: Vec<i64> = Vec::new();
    for line in input {
        let split = line.split_whitespace().collect::<Vec<&str>>();
        first_column.push(split[0].parse::<i64>().unwrap());
        second_column.push(split[1].parse::<i64>().unwrap());
    }

    [first_column, second_column]
}

fn part1(input: [Vec<i64>; 2]) -> i64 {
    let mut first_column = input[0].clone();
    let mut second_column = input[1].clone();
    first_column.sort();
    second_column.sort();
    let mut result = 0;
    for i in 0..first_column.len() {
        result += (first_column[i] - second_column[i]).abs();
    }
    result
}

fn part2(input: [Vec<i64>; 2]) -> usize{
    let mut result = 0;
    for i in 0..input[0].len() {
        let count = input[1].clone().into_iter().filter(|&x| x == input[0][i]).count();
        result += input[0][i] as usize * count;
    }
    result
}

fn main(){
    let input = read_input(args().nth(1).unwrap());
    println!("Input: --------");
    println!("{:?}", input);
    println!("The answer to part 1 is: {}", part1(split_string(input.clone())));
    println!("The answer to part 2 is: {}", part2(split_string(input)));
}

File: go/day1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
	"math"
)

func readFile(filename string) [2][]int {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	var fileLines [2][]int
	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		var pairs []int
		for _, field := range fields {
			num, err := strconv.Atoi(field)
			if err != nil {
				panic(err)
			}
			pairs = append(pairs, num)
		}
		fileLines[0] = append(fileLines[0], pairs[0])
		fileLines[1] = append(fileLines[1], pairs[1])
	}
	file.Close()
	return fileLines
}

func part1(fileLines [2][]int) int {
	fmt.Println(fileLines)
	first_column := fileLines[:][0]
	second_column := fileLines[:][1]
	slices.Sort(first_column)
	slices.Sort(second_column)
	var result int = 0
	for i, num1 := range first_column {
		result += int(math.Abs(float64(num1 - second_column[i])))
	}
	return result
}

func part2(fileLines [2][]int) int {
	var result int = 0
	freq := make(map[int]int)
	for _, num := range fileLines[1] {
		freq[num]++
	}
	for _, num := range fileLines[0] {
		if freq[num] > 0 {
			result += num*freq[num]
		}
	}
	return result
}

func main() {

	var filename string = os.Args[1]

	// Read file
	fileLines := readFile(filename)

	// Print the content of the file
	fmt.Println("The content of the file is:")
	for _, line := range fileLines {
		fmt.Println(line)
	}

	// Print answer for part 1
	fmt.Println("The answer for part 1 is: ", part1(fileLines))

	// Print answer for part 2
	fmt.Println("The answer for part 2 is: ", part2(fileLines))
	
}